### PR TITLE
Symmetric Schur and Simultaneous Diagonalization

### PR DESCRIPTION
**Pull request by jeremylt**
This ~WIP~ PR adds Symmetric QR and Simultaneous Diagonalization.

This is step 1 of 4 for the FDM preconditioner.
1) Simultaneous Diagonalization
2) QFunction Gallery
3) Backend diagonal setup
4) FDM (use QF gallery from 2 to create operator using basis from 1 and qdata from 3)

3 should be relatively lightweight, so backends could possibly offload 3 to ref and rearrange the vector as desired.

### CODE REVIEW

File: interface/ceed-basis.c [hunk 1]
```diff
@@ -540,6 +540,41 @@ int CeedQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *tau,
...
+
+  @ref Utility
+**/
+int CeedSymmetricQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *lambda,
+                                 CeedInt n) {
+  return CeedError(ceed, 1, "Not implemented");
+}
+
+/**
+  @brief Return Simultaneous Diagonalization of two matrices
```
Comment (jedbrown) [inline interface/ceed-basis.c]: Description should specify necessary predicates and possibly be described in terms of generalized eigenvalue problems.

File: interface/ceed-basis.c [hunk 2]
```diff
@@ -540,6 +540,41 @@ int CeedQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *tau,
   return 0;
 }
 
+/**
+  @brief Return Symmetric QR Factorization of matrix
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-basis.c]: Yes, technically I need the Schur decomposition and I am using symmetric QR to do so. It seems to be the preferred option in Golub and Van Loan for our size matrices.

File: interface/ceed-basis.c [hunk 3]
```diff
@@ -540,6 +540,41 @@ int CeedQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *tau,
...
+
+  @param[out] mat     Row-major matrix to be factorized in place
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-basis.c]: I'll update the current as well.

File: interface/ceed-basis.c [hunk 4]
```diff
@@ -426,18 +426,18 @@ int CeedBasisView(CeedBasis basis, FILE *stream) {
...
     where A is an mxn matrix indexed as A[i*row + j*col]
 
-  @param[out] A  Matrix to apply Householder reflection to, in place
-  @param v       Householder vector
-  @param b       Scaling factor
-  @param m       Number of rows in A
-  @param n       Number of columns in A
-  @param row     Col stride
-  @param col     Row stride
+  @param[in/out] A  Matrix to apply Householder reflection to, in place
```
Comment (jedbrown) [inline interface/ceed-basis.c]: > Possible values are "[in]", "[in,out]", and "[out]"
http://www.doxygen.nl/manual/commands.html#cmdparam

**Comment by jeremylt** [comment]
The Schur decomp works; I just need to use it to set up the simultaneous diagonalization.

**Comment by jeremylt** [comment]
@jedbrown, this should be good to go now. I used the mass and diffusion matrices as my test cases. I also reordered the basis tests to group similar tests (t30* are utility functions, t31* are tensor bases, t32* are simplex bases)

**Comment by jeremylt** [comment]
Rebased for #290.

**Comment by jeremylt** [comment]
I added the equivalent LAPACK function, sygv.

File: tests/t316-basis.c [hunk 5]
```diff
@@ -0,0 +1,89 @@
+/// @file
+/// Test inderintegrated grad in multiple dimensions
+/// \test Test inderintegraded grad in multiple dimensions
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply tests/t316-basis.c]: Words are hard. Fixed.

File: interface/ceed-basis.c [hunk 6]
```diff
@@ -526,20 +561,234 @@ int CeedQRFactorization(Ceed ceed, CeedScalar *mat, CeedScalar *tau,
...
+  @param[out] matC    Row-major output matrix C
+  @param m            Number of rows of C
+  @param n            Number of columns of C
+  @param kk           Number of columns of A/rows of B
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Utility
+**/
+int CeedMatrixMultiply(Ceed ceed, CeedScalar *matA, CeedScalar *matB,
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-basis.c]: Fixed. I got right on `CeedGivensRotation` but forgot it here.

**Comment by jeremylt** [comment]
Rebasing would be a mess, so I merged master into here.

**Comment by jedbrown** [comment]
You can squash-merge or squash and split the resulting single commit to separate test renaming from the new feature.

**Comment by jeremylt** [comment]
Squash-merge would be easier. I'd have to pull the commits apart to sort out the changes from master cleanly.

**Comment by jeremylt** [comment]
@jedbrown, your comment seems to indicate that you approve the merge, but it doesn't seem to have updated in GitHub.

File: interface/ceed-basis.c [hunk 7]
```diff
@@ -497,33 +497,68 @@ static int CeedHouseholderApplyQ(CeedScalar *A, const CeedScalar *Q,
   return 0;
 }
 
+/**
+  @brief Compute Givens rotation
+
+    Computes A = G A (or G^T A in transpose mode)
+    where A is an mxn matrix indexed as A[i*row + j*col]
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-basis.c]: fixed

File: interface/ceed-basis.c [hunk 8]
```diff
@@ -467,15 +467,15 @@ static int CeedHouseholderReflect(CeedScalar *A, const CeedScalar *v,
 
     Compute A = Q A where Q is mxk and A is mxn. k<m
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-basis.c]: Fixed: 'Compute A = Q A where Q is mxm and A is mxn. Q targets row k, where k<m.'
That was bad documentation.

[the remark this answers is not in the document]
Comment (jedbrown) [reply interface/ceed-basis.c]: What do you mean "targets row k"? Even with `k=1` vector, `I - v v^T` is generally a dense matrix so it's using information from the entire matrix and updating the entire matrix.

[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-basis.c]: Ah, that's where the `k` in `mxk` came from, the number of reflectors as Q is stored in factored form. 13 month ago me wrote bad documentation for the code me/us of 16 months ago wrote.

**Pull request merged**